File: framework/tests/st/machine/src/ops/test_dynamic_bmm.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------

""" 相关用例 Golden 生成逻辑.
本脚本有 2 种执行模式:
1. CI批跑时, 由 cmake/scripts/golden_ctrl.py 调用, 为避免日志过多, 此时 logging 级别为 logging.INFO;
2. 单独调用时, 本脚本单独被调用, 此时 logging 级别为 logging.DEBUG;
"""

import math
import sys
import logging
from pathlib import Path

import numpy as np
from ml_dtypes import bfloat16

if __name__ == "__main__":
    # 系统 import 路径
    global_src_root: Path = Path(Path(__file__).parent, "../../../../../").resolve()
    # 日志级别
    logging.basicConfig(format='%(asctime)s - %(filename)s:%(lineno)d - %(levelname)s: %(message)s',
                        level=logging.DEBUG)
    logging.debug("SrcRoot: %s", global_src_root)
    g_ctrl_path: Path = Path(global_src_root, "tests/cmake/scripts")
    if str(g_ctrl_path) not in sys.path:
        sys.path.append(str(g_ctrl_path))
    from golden_register import GoldenRegister
else:
    from golden_register import GoldenRegister

BF16 = bfloat16
INT32 = np.int32
FP32 = np.float32
FP16 = np.float16
INT8 = np.int8


def ceil_div(a, b):
    return (a + b - 1) // b


def gen_axes_for_transpose(offset, base):
    return [x for x in range(offset)] + [x + offset for x in base]


def nd_to_fractal_nz(data: np.ndarray):
    ori_shape = data.shape
    batch_ori = ori_shape[:-2]
    m_ori, n_ori = ori_shape[-2:]
    batch_num = len(batch_ori)
    batch_padding = ((0, 0),) * batch_num
    if data.dtype == FP16 or data.dtype == BF16 or data.dtype == INT32:
        m0, n0 = 16, 16
    elif data.dtype == FP32:
        m0, n0 = 16, 8
    elif data.dtype == INT8:
        m0, n0 = 16, 32
    m1, n1 = ceil_div(m_ori, m0), ceil_div(n_ori, n0)
    padding_n = n1 * n0 - n_ori
    padding_m = m1 * m0 - m_ori
    data = np.pad(data, (batch_padding + ((0, padding_m), (0, padding_n))), 'constant')
    array_trans = gen_axes_for_transpose(len(data.shape) - 2, [2, 0, 1, 3])
    data = data.reshape(batch_ori + (m1, m0, n1, n0)).transpose(*array_trans)
    return data


class ShapeConfigOneBatch:
    def __init__(self, b: int, m: int, k: int, n: int, in_dtype: np.dtype, out_dtype: np.dtype, trans_a: bool,
                 trans_b: bool, a_nz_flag: bool, b_nz_flag: bool, c_nz_flag: bool):
        self.b = b
        self.m = m
        self.k = k
        self.n = n
        self.trans_a = trans_a
        self.trans_b = trans_b
        self.in_dtype = in_dtype
        self.out_dtype = out_dtype
        self.a_nz_flag = a_nz_flag
        self.c_nz_flag = c_nz_flag
        self.b_nz_flag = b_nz_flag


class ShapeConfigTwoBatch:
    def __init__(self, b1: int, b2: int, m: int, k: int, n: int, in_dtype: np.dtype, out_dtype: np.dtype, trans_a: bool,
                 trans_b: bool, a_nz_flag: bool, b_nz_flag: bool, c_nz_flag: bool):
        self.b1 = b1
        self.b2 = b2
        self.n = n
        self.m = m
        self.k = k
        self.in_dtype = in_dtype
        self.out_dtype = out_dtype
        self.a_nz_flag = a_nz_flag
        self.b_nz_flag = b_nz_flag
        self.c_nz_flag = c_nz_flag
        self.trans_b = trans_b
        self.trans_a = trans_a


def gen_bmm_data(input_config: ShapeConfigOneBatch, output_dir: Path):
    shape_a = [input_config.b, input_config.m, input_config.k]
    shape_b = [input_config.b, input_config.k, input_config.n]
    shape_c = [input_config.b, input_config.m, input_config.n]

    c_path = Path(output_dir, 'mat_c.bin')
    a_path = Path(output_dir, 'mat_a.bin')
    b_path = Path(output_dir, 'mat_b.bin')

    if input_config.in_dtype == FP16:
        a = np.random.uniform(-1, 1, shape_a).astype(FP16)
        b = np.random.uniform(-1, 1, shape_b).astype(FP16)
        c = np.matmul(a.astype(FP32), b.astype(FP32))
    elif input_config.in_dtype == INT8:
        a = np.random.randint(-4, 5, shape_a).astype(INT8)
        b = np.random.randint(-4, 5, shape_b).astype(INT8)
        c = np.matmul(a.astype(INT32), b.astype(INT32)).astype(INT32)
    elif input_config.in_dtype == FP32:
        a = np.random.uniform(-1, 1, shape_a).astype(FP32)
        b = np.random.uniform(-1, 1, shape_b).astype(FP32)
        c = np.matmul(a.astype(FP32), b.astype(FP32))
    elif input_config.in_dtype == BF16:
        a = np.random.uniform(-1, 1, shape_a).astype(BF16)
        b = np.random.uniform(-1, 1, shape_b).astype(BF16)
        c = np.matmul(a.astype(FP32), b.astype(FP32))
    c = c.astype(input_config.out_dtype)

    if input_config.trans_a:
        a = a.transpose(0, 2, 1)
    if input_config.a_nz_flag:
        a = nd_to_fractal_nz(a)

    if input_config.trans_b:
        b = b.transpose(0, 2, 1)
    if input_config.b_nz_flag:
        b = nd_to_fractal_nz(b)
    a.tofile(a_path)
    b.tofile(b_path)

    if input_config.c_nz_flag:
        c = nd_to_fractal_nz(c)
    c.tofile(c_path)


def gen_bmm_data_two_batch(input_config: ShapeConfigTwoBatch, output_dir: Path):
    shape_a = [input_config.b1, input_config.b2, input_config.m, input_config.k]
    shape_b = [input_config.b1, input_config.b2, input_config.k, input_config.n]
    shape_c = [input_config.b1, input_config.b2, input_config.m, input_config.n]

    a_path = Path(output_dir, 'mat_a.bin')
    c_path = Path(output_dir, 'mat_c.bin')
    b_path = Path(output_dir, 'mat_b.bin')

    if input_config.in_dtype == FP16:
        b = np.random.uniform(-1, 1, shape_b).astype(FP16)
        a = np.random.uniform(-1, 1, shape_a).astype(FP16)
        c = np.matmul(a.astype(FP32), b.astype(FP32))
    elif input_config.in_dtype == BF16:
        b = np.random.uniform(-1, 1, shape_b).astype(BF16)
        a = np.random.uniform(-1, 1, shape_a).astype(BF16)
        c = np.matmul(a.astype(FP32), b.astype(FP32))
    elif input_config.in_dtype == FP32:
        a = np.random.uniform(-1, 1, shape_a).astype(FP32)
        b = np.random.uniform(-1, 1, shape_b).astype(FP32)
        c = np.matmul(a.astype(FP32), b.astype(FP32))
    elif input_config.in_dtype == INT8:
        a = np.random.randint(-4, 5, shape_a).astype(INT8)
        b = np.random.randint(-4, 5, shape_b).astype(INT8)
        c = np.matmul(a.astype(INT32), b.astype(INT32)).astype(INT32)

    if input_config.trans_a:
        a = a.swapaxes(2, 3)
    if input_config.a_nz_flag:
        a = nd_to_fractal_nz(a)

    if input_config.trans_b:
        b = a.swapaxes(2, 3)
    if input_config.b_nz_flag:
        b = nd_to_fractal_nz(b)
    a.tofile(a_path)
    b.tofile(b_path)

    if input_config.c_nz_flag:
        c = nd_to_fractal_nz(c)
    c.tofile(c_path)


@GoldenRegister.reg_golden_func(
    case_names=[
        "DynamicBatchMatmulTest.test_bmm_A_Bt_ND_fp16",
        "DynamicBatchMatmulTest.test_bmm_A_Bt_NZ_fp16",
        "DynamicBatchMatmulTest.test_bmm_A_B_ND_bf16_tile1",
        "DynamicBatchMatmulTest.test_bmm_At_Bt_ND_fp16",
        "DynamicBatchMatmulTest.test_bmm_At_Bt_ANZ_BND_fp16",
    ]
)
def gen_dynamic_bmm_golden(case_name: str, output: Path) -> bool:
    if case_name == "DynamicBatchMatmulTest.test_bmm_A_Bt_ND_fp16":
        input_config = ShapeConfigOneBatch(3, 2, 576, 4096, FP16, FP32, False, True, False, False, False)
        gen_bmm_data(input_config, output)
        return True
    if case_name == "DynamicBatchMatmulTest.test_bmm_A_Bt_NZ_fp16":
        input_config = ShapeConfigOneBatch(4, 96, 128, 256, FP16, FP32, False, True, False, True, False)
        gen_bmm_data(input_config, output)
        return True
    if case_name == "DynamicBatchMatmulTest.test_bmm_A_B_ND_bf16_tile1":
        input_config = ShapeConfigOneBatch(6, 1, 576, 4096, BF16, FP32, False, False, False, False, False)
        gen_bmm_data(input_config, output)
        return True
    if case_name == "DynamicBatchMatmulTest.test_bmm_At_Bt_ND_fp16":
        input_config = ShapeConfigOneBatch(3, 2, 576, 4096, FP16, FP32, True, True, False, False, False)
        gen_bmm_data(input_config, output)
        return True
    if case_name == "DynamicBatchMatmulTest.test_bmm_At_Bt_ANZ_BND_fp16":
        input_config = ShapeConfigOneBatch(3, 128, 256, 512, FP16, FP32, True, True, True, False, False)
        gen_bmm_data(input_config, output)
        return True
    else:
        logging.error("Can't get func to gen golden, case(%s)", case_name)
        return False

File: framework/tests/st/machine/src/ops/test_dynamic_bmm.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
 * \file test_dynamic_bmm.cpp
 * \brief
 */

#include <gtest/gtest.h>
#include "tilefwk/data_type.h"
#include "interface/function/function.h"
#include "tilefwk/tilefwk_op.h"
#include "test_suite_stest_ops.h"
#include "interface/interpreter/raw_tensor_data.h"
#include "test_dev_func_runner.h"

using namespace npu::tile_fwk;
using namespace npu::tile_fwk::dynamic;

namespace {
const size_t BMM_SHAPE_SIZE = 4;
const size_t BMM_VIEW_SHAPE_SIZE = 3;
const size_t BMM_SHAPE_N_DIM = 2;
const size_t BMM_SHAPE_N_IDX = 3;
class DynamicBatchMatmulTest : public npu::tile_fwk::stest::TestSuite_STest_Ops_Aihac {};

template <typename dtype>
Tensor constructMatmulTensor(const std::vector<int64_t>& shape, const string& name, bool isNz)
{
    auto dataType = GetAstDtype<dtype>();
    return isNz ? Tensor(dataType, shape, name, TileOpFormat::TILEOP_NZ) : Tensor(dataType, shape, name);
}

inline SymbolicScalar CeilDivSymbolicScalar(SymbolicScalar a, int64_t b) { return b == 0 ? a : (a + b - 1) / b; }

template <typename outputDtype, bool transA, bool transB, bool isCNz>
static void NonSplitFunc(const Tensor& tensor_a, const Tensor& tensor_b, Tensor& tensor_c)
{
    const auto& aShape = tensor_a.GetShape();
    std::vector<SymbolicScalar> aValidShape = {aShape[0], aShape[1], aShape[2]};
    const auto& bShape = tensor_b.GetShape();
    std::vector<SymbolicScalar> bValidShape = {bShape[0], bShape[1], bShape[2]};

    int64_t m = transA ? aShape[2] : aShape[1];
    int64_t k = transA ? aShape[1] : aShape[2];
    int64_t n = transB ? bShape[1] : bShape[2];

    FUNCTION("testNoSplit", {tensor_a, tensor_b}, {tensor_c})
    {
        LOOP("mLoop", FunctionType::DYNAMIC_LOOP, mIdx, LoopRange(1))
        {
            Tensor dyn_a = View(tensor_a, aShape, aValidShape, {0, mIdx, 0});
            Tensor dyn_b = View(tensor_b, bShape, bValidShape, {0, 0, 0});
            TileShape::Current().SetMatrixSize({m, k, n});
            tensor_c = Matrix::BatchMatmul(GetAstDtype<outputDtype>(), dyn_a, dyn_b, transA, transB, isCNz);
        }
    }
}

template <typename outputDtype, bool transA, bool transB, bool isCNz>
static void MSplitFunc(
    const std::vector<int64_t>& viewShape, const Tensor& tensor_a, const Tensor& tensor_b, Tensor& tensor_c)
{
    const auto& aShape = tensor_a.GetShape();
    std::vector<SymbolicScalar> aValidShape = {aShape[0], aShape[1], aShape[2]};
    const auto& bShape = tensor_b.GetShape();
    std::vector<SymbolicScalar> bValidShape = {bShape[0], bShape[1], bShape[2]};

    int64_t n = transB ? bShape[1] : bShape[2];
    int64_t m = transA ? aShape[2] : aShape[1];
    int64_t k = transA ? aShape[1] : aShape[2];

    FUNCTION("testMSplit", {tensor_a, tensor_b}, {tensor_c})
    {
        LOOP(
            "mLoop", FunctionType::DYNAMIC_LOOP, mIdx,
            LoopRange(0, CeilDivSymbolicScalar(transA ? aShape[BMM_SHAPE_N_DIM] : aShape[1], viewShape[1]), 1))
        {
            Tensor dyn_a;
            if (transA) {
                dyn_a = View(
                    tensor_a, {aShape[0], aShape[1], viewShape[1]},
                    {aShape[0], aShape[1], std::min(aShape[2] - viewShape[1] * mIdx, viewShape[1])},
                    {0, 0, mIdx * viewShape[1]});
            } else {
                dyn_a = View(
                    tensor_a, {aShape[0], viewShape[1], aShape[2]},
                    {aShape[0], std::min(aShape[1] - viewShape[1] * mIdx, viewShape[1]), aShape[2]},
                    {0, mIdx * viewShape[1], 0});
            }
            Tensor dyn_b = View(tensor_b, bShape, bValidShape, {0, 0, 0});
            TileShape::Current().SetMatrixSize({m, k, n});
            Tensor res = Matrix::BatchMatmul(GetAstDtype<outputDtype>(), dyn_a, dyn_b, transA, transB, isCNz);
            Assemble(res, {0, mIdx * viewShape[0], 0}, tensor_c);
        }
    }
}

template <typename outputDtype, bool transA, bool transB, bool isCNz>
static void NSplitFunc(
    const std::vector<int64_t>& viewShape, const Tensor& tensor_a, const Tensor& tensor_b, Tensor& tensor_c)
{
    const auto& aShape = tensor_a.GetShape();
    std::vector<SymbolicScalar> aValidShape = {aShape[0], aShape[1], aShape[2]};
    const auto& bShape = tensor_b.GetShape();
    std::vector<SymbolicScalar> bValidShape = {bShape[0], bShape[1], bShape[2]};

    int64_t m = transA ? aShape[2] : aShape[1];
    int64_t k = transA ? aShape[1] : aShape[2];
    int64_t n = transB ? bShape[1] : bShape[2];

    FUNCTION("testNSplit", {tensor_a, tensor_b}, {tensor_c})
    {
        LOOP(
            "nLoop", FunctionType::DYNAMIC_LOOP, nIdx,
            LoopRange(
                0, CeilDivSymbolicScalar(transB ? bShape[1] : bShape[BMM_SHAPE_N_DIM], viewShape[BMM_SHAPE_N_DIM]), 1))
        {
            Tensor dyn_a = View(tensor_a, aShape, aValidShape, {0, 0, 0});
            Tensor dyn_b;
            if (!transB) {
                dyn_b = View(
                    tensor_b, {bShape[0], bShape[1], viewShape[2]},
                    {bShape[0], bShape[1], std::min(bShape[2] - viewShape[2] * nIdx, viewShape[2])},
                    {0, 0, nIdx * viewShape[2]});
            } else {
                dyn_b = View(
                    tensor_b, {bShape[0], viewShape[2], bShape[2]},
                    {bShape[0], std::min(bShape[1] - viewShape[2] * nIdx, viewShape[2]), bShape[2]},
                    {0, nIdx * viewShape[2], 0});
            }
            TileShape::Current().SetMatrixSize({m, k, n});
            Tensor res = Matrix::BatchMatmul(GetAstDtype<outputDtype>(), dyn_a, dyn_b, transA, transB, isCNz);
            Assemble(res, {0, 0, nIdx * viewShape[2]}, tensor_c);
        }
    }
}

template <typename outputDtype, bool transA, bool transB, bool isCNz>
static void MNSplitFunc(
    const std::vector<int64_t>& viewShape, const Tensor& tensor_a, const Tensor& tensor_b, Tensor& tensor_c)
{
    const auto& aShape = tensor_a.GetShape();
    std::vector<SymbolicScalar> aValidShape = {aShape[0], aShape[1], aShape[2]};
    const auto& bShape = tensor_b.GetShape();
    std::vector<SymbolicScalar> bValidShape = {bShape[0], bShape[1], bShape[2]};

    int64_t m = transA ? aShape[2] : aShape[1];
    int64_t n = transB ? bShape[1] : bShape[2];
    int64_t k = transA ? aShape[1] : aShape[2];

    FUNCTION("testNSplit", {tensor_a, tensor_b}, {tensor_c})
    {
        LOOP(
            "mLoop", FunctionType::DYNAMIC_LOOP, mIdx,
            LoopRange(0, CeilDivSymbolicScalar(transA ? aShape[BMM_SHAPE_N_DIM] : aShape[1], viewShape[1]), 1))
        {
            LOOP(
                "nLoop", FunctionType::DYNAMIC_LOOP, nIdx,
                LoopRange(
                    0, CeilDivSymbolicScalar(transB ? bShape[1] : bShape[BMM_SHAPE_N_DIM], viewShape[BMM_SHAPE_N_DIM]),
                    1))
            {
                Tensor dyn_a;
                if (transA) {
                    dyn_a = View(
                        tensor_a, {aShape[0], aShape[1], viewShape[1]},
                        {aShape[0], aShape[1], std::min(aShape[2] - viewShape[1] * mIdx, viewShape[1])},
                        {0, 0, mIdx * viewShape[1]});
                } else {
                    dyn_a = View(
                        tensor_a, {aShape[0], viewShape[1], aShape[2]},
                        {aShape[0], std::min(aShape[1] - viewShape[1] * mIdx, viewShape[1]), aShape[2]},
                        {0, mIdx * viewShape[1], 0});
                }
                Tensor dyn_b;
                if (!transB) {
                    dyn_b = View(
                        tensor_b, {bShape[0], bShape[1], viewShape[2]},
                        {bShape[0], bShape[1], std::min(bShape[2] - viewShape[2] * nIdx, viewShape[2])},
                        {0, 0, nIdx * viewShape[2]});
                } else {
                    dyn_b = View(
                        tensor_b, {bShape[0], viewShape[2], bShape[2]},
                        {bShape[0], std::min(bShape[1] - viewShape[2] * nIdx, viewShape[2]), bShape[2]},
                        {0, nIdx * viewShape[2], 0});
                }
                TileShape::Current().SetMatrixSize({m, k, n});
                Tensor res = Matrix::BatchMatmul(GetAstDtype<outputDtype>(), dyn_a, dyn_b, transA, transB, isCNz);
                Assemble(res, {0, mIdx * viewShape[0], nIdx * viewShape[1]}, tensor_c);
            }
        }
    }
}

template <typename inputDtype, typename outputDtype, bool transA, bool transB, bool isCNz>
void TestDynBatchMatmul(
    const std::vector<int64_t>& mmShape, bool isANz, bool isBNz, const std::vector<int64_t>& viewShape, string dataPath)
{
    SetInterpreterConfig();

    if (mmShape.size() != BMM_SHAPE_SIZE || viewShape.size() != BMM_VIEW_SHAPE_SIZE) {
        return;
    }
    int64_t b = mmShape[0];
    int64_t m = mmShape[1];
    int64_t k = mmShape[2];
    int64_t n = mmShape[BMM_SHAPE_N_IDX];
    Tensor tensor_a = transA ? constructMatmulTensor<inputDtype>({b, k, m}, "tensor_a", isANz) :
                               constructMatmulTensor<inputDtype>({b, m, k}, "tensor_a", isANz);
    Tensor tensor_b = transB ? constructMatmulTensor<inputDtype>({b, n, k}, "tensor_b", isBNz) :
                               constructMatmulTensor<inputDtype>({b, k, n}, "tensor_b", isBNz);
    Tensor tensor_c = constructMatmulTensor<outputDtype>({b, m, n}, "tensor_c", isCNz);

    std::vector<inputDtype> aData(b * m * k, 0);
    std::vector<inputDtype> bData(b * k * n, 0);
    std::vector<outputDtype> golden(b * m * n, 0);

    readInput<outputDtype>(dataPath + "/mat_c.bin", golden);
    readInput<inputDtype>(dataPath + "/mat_a.bin", aData);
    readInput<inputDtype>(dataPath + "/mat_b.bin", bData);

    ProgramData::GetInstance().AppendOutputs({
        RawTensorData::CreateConstantTensor<outputDtype>(tensor_c, 0.0f),
    });

    ProgramData::GetInstance().AppendInputs({
        RawTensorData::CreateTensor<inputDtype>(tensor_a, aData),
        RawTensorData::CreateTensor<inputDtype>(tensor_b, bData),
    });

    ProgramData::GetInstance().AppendGoldens({
        RawTensorData::CreateTensor<outputDtype>(tensor_c, golden),
    });

    int64_t viewM = viewShape[1];
    int64_t viewN = viewShape[2];
    if (viewM > 0 && viewN > 0) {
        MNSplitFunc<outputDtype, transA, transB, isCNz>(viewShape, tensor_a, tensor_b, tensor_c);
    } else if (viewM > 0) {
        MSplitFunc<outputDtype, transA, transB, isCNz>(viewShape, tensor_a, tensor_b, tensor_c);
    } else if (viewN > 0) {
        NSplitFunc<outputDtype, transA, transB, isCNz>(viewShape, tensor_a, tensor_b, tensor_c);
    } else {
        NonSplitFunc<outputDtype, transA, transB, isCNz>(tensor_a, tensor_b, tensor_c);
    }

    // excute
    DevFuncRunner::Run(Program::GetInstance().GetLastFunction());
    auto outs = npu::tile_fwk::ProgramData::GetInstance().GetOutputData(0);
    EXPECT_TRUE(resultCmp(golden, (outputDtype*)outs->data(), 0.001f));
}

TEST_F(DynamicBatchMatmulTest, test_bmm_A_Bt_ND_fp16)
{
    TileShape::Current().SetCubeTile({128, 128}, {128, 128}, {128, 128});
    int64_t b = 3;
    int64_t m = 2;
    int64_t k = 576;
    int64_t n = 4096;
    bool isANz = false;
    bool isBNz = false;
    std::vector<int64_t> viewShape = {-1, -1, -1};
    TestDynBatchMatmul<npu::tile_fwk::float16, float, false, true, false>(
        {b, m, k, n}, isANz, isBNz, viewShape, GetGoldenDir());
}

TEST_F(DynamicBatchMatmulTest, test_bmm_A_Bt_NZ_fp16)
{
    TileShape::Current().SetCubeTile({128, 128}, {128, 128}, {128, 128});
    int64_t b = 4;
    int64_t m = 96;
    int64_t k = 128;
    int64_t n = 256;
    bool isANz = false;
    bool isBNz = true;
    std::vector<int64_t> viewShape = {-1, -1, -1};
    TestDynBatchMatmul<npu::tile_fwk::float16, float, false, true, false>(
        {b, m, k, n}, isANz, isBNz, viewShape, GetGoldenDir());
}

TEST_F(DynamicBatchMatmulTest, test_bmm_A_B_ND_bf16_tile1)
{
    TileShape::Current().SetCubeTile({128, 128}, {256, 256}, {128, 128});
    int64_t b = 6;
    int64_t m = 1;
    int64_t k = 576;
    int64_t n = 4096;
    bool isANz = false;
    bool isBNz = false;
    std::vector<int64_t> viewShape = {-1, -1, -1};
    TestDynBatchMatmul<npu::tile_fwk::bfloat16, float, false, false, false>(
        {b, m, k, n}, isANz, isBNz, viewShape, GetGoldenDir());
}

TEST_F(DynamicBatchMatmulTest, test_bmm_At_Bt_ND_fp16)
{
    TileShape::Current().SetCubeTile({128, 128}, {128, 128}, {128, 128});
    int64_t b = 3;
    int64_t m = 2;
    int64_t k = 576;
    int64_t n = 4096;
    bool isANz = false;
    bool isBNz = false;
    std::vector<int64_t> viewShape = {-1, -1, -1};
    TestDynBatchMatmul<npu::tile_fwk::float16, float, true, true, false>(
        {b, m, k, n}, isANz, isBNz, viewShape, GetGoldenDir());
}

TEST_F(DynamicBatchMatmulTest, test_bmm_At_Bt_ANZ_BND_fp16)
{
    TileShape::Current().SetCubeTile({128, 128}, {128, 128}, {128, 128});
    int64_t b = 3;
    int64_t m = 128;
    int64_t k = 256;
    int64_t n = 512;
    bool isANz = true;
    bool isBNz = false;
    std::vector<int64_t> viewShape = {-1, -1, -1};
    TestDynBatchMatmul<npu::tile_fwk::float16, float, true, true, false>(
        {b, m, k, n}, isANz, isBNz, viewShape, GetGoldenDir());
}
} // namespace
